### PR TITLE
Add Cloud Services UI

### PR DESCRIPTION
**Pull request by PanSpagetka**
Add Cloud Services UI. 

Known issues:
-  Buttons are not working (API is not done yet)
- <s>Type should be in more human readable form</s>
- `Red Hat Cloud Services` does not fit into menu, can we have something shorter?
- No redirect to Cloud Services

Screenshot:
Screenshot from 2019-06-27 11-06-21

ping @martinpovolny @h-kataria @Hyperkid123

### CODE REVIEW

**Comment by terezanovotna** [comment]
- for Red Hat Cloud Services in the menu, can we have Red Hat new logo as an icon?
- we can abbreviate it to RH Cloud Services for now
 
For type, how about dividing it into two columns? (dropping  words ManageIQ and Providers)
- Provider: Amazon, Ansible Tower, Azure etc. 
- Type: Cloud Manager, Storage Manager, Network Manager etc.

**Comment by terezanovotna** [comment]
What happens when the user clicks on the synchronize? can we show some kind of status info? it would be helpful to show that.

File: app/javascript/packs/helper.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    let result = [];
+    if (showIcon) {
+      result = [{
+        propery: 'icon',
+        header: {
+          label: '',
+          formatters: [headerFormat],
+        },
+      },
+      ...result];
```
Comment (Hyperkid123) [inline app/javascript/packs/helper.js]: `...result` is not necessary here, there is nothing to add.

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    property: 'type'
+  },
+  {
+    label: 'Action',
+    property: 'action'
+  },
+];
+
+
+const RedHatCloudServices = (_props) => {
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: if you don't intend to use props just don't access the argument `const Foo = () => {...}`

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        dispatch({type: 'setRows', rows: orderBy(rows, 'name', 'asc')})
+      });
+    }, [])
+
+    useEffect(() => {
+      setPage()(pagination.page);
+    }, [rows])
+
+    useEffect(() => {
+      dispatch({type: 'setTotal', total: rows.filter(row => row[currentFilterType.id].toLowerCase().includes(currentValue.toLowerCase())).length});
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: i would consider using `localeCompare` i think it may improve filtering capabilities: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }, [pagination])
+
+    useEffect(() => {
+      dispatch({type: 'setTotal', total: rows.filter(row => row[currentFilterType.id].toLowerCase().includes(currentValue.toLowerCase())).length});
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: ^^

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import { TablePfProvider, Table, Filter, FormControl, Button, Checkbox, Paginator, PAGINATION_VIEW } from 'patternfly-react'
+import orderBy from 'lodash/orderBy'
+import RedHatCloudServicesTable from './red_hat_cloud_services_table';
+import { reducer, selectRow } from './reducers';
+import {  setPageWrapper, paginate } from './helper'
+
+const filterFields = [
+  {
+    id: 'name',
+    title: 'Name',
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: gettext for string constants :wink:

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+          <form className="toolbar-pf-actions">
+            <div className="form-group toolbar-pf-filter">
+              <Filter>
+                <Filter.TypeSelector
+                  filterTypes={filterFields}
+                  currentFilterType={currentFilterType}
+                  onFilterTypeSelected={(filterType) => {
+                    if (currentFilterType !== filterType) {
+                      dispatch({ type: 'setFilterValue', value: ''});
+                      dispatch({ type: 'setFilterType', filterType: filterType});
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: You are doing multiple state updates which will result in multiple render cycles. I would consider creating new action which will do both state mutations at once.

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                    }
+                  }}
+                />
+                <FormControl
+                  type={currentFilterType.filterType}
+                  value={currentValue}
+                  placeholder={currentFilterType.placeholder}
+                  onChange={({ target: { value: filterValue } }) => {
+                    // setCurrentValue(filterValue);
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: comment

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                    dispatch({ type: 'setFilterValue', value: filterValue});
+                  }}
+                />
+              </Filter>
+            </div>
+            <div class="form-group">
+              <button class="btn btn-default" type="button" id="upload-selected">Upload</button>
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: gettext

File: app/javascript/packs/red_hat_cloud_services.jsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+              <button class="btn btn-default" type="button" id="Synchronize" disabled>Synchronize</button>
```
Comment (Hyperkid123) [inline app/javascript/packs/red_hat_cloud_services.jsx]: This is always disabled. Is it intentional?

File: app/javascript/packs/reducers.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    case 'sortColumn':
+      return {...state, ...sortColumn(state, action)};
+    case 'selectRow':
+      return {...state, ...selectRow(state, action)};
+    case 'setPagination':
+      return {...state, pagination: action.pagination};
+    case 'setTotal':
+      return {...state, total: action.total}
+    default:
+      return state;
```
Comment (Hyperkid123) [inline app/javascript/packs/reducers.js]: Since this is not global state and it does not listen to any actions apart those that are called via directly connected dispatch, common pattern is throwing an error if you have unknown action type.

**Comment by PanSpagetka** [comment]
> What happens when the user clicks on the synchronize? can we show some kind of status info? it would be helpful to show that.

I have discussed with @lpichler and if I understand it correctly it will initiate task that will upload data from ManageIQ to Cloud Services. That could take some time so there is no immediate feedback. I could add "Synchronization task was initiated." flash message, but that's it.

File: app/javascript/packs/red_hat_cloud_services.jsx (hunk 3 above)
Comment (Hyperkid123) [reply app/javascript/packs/red_hat_cloud_services.jsx]: nvm you are filtering not matching strings

**Comment by Fryguy** [comment]
@terezanovotna Previously we had the Insights left-side menu as the second from the top (under Cloud Intel).  Do we still want it there?

**Comment by Fryguy** [comment]
>  but considering we are kind of in hurry

Are we?

**Comment by Fryguy** [comment]
> or Red Hat Cloud Services in the menu, can we have Red Hat new logo as an icon?
> we can abbreviate it to RH Cloud Services for now

If we are going to have the logo, then do we need "RH" at all?  Perhaps just "Cloud Services"?

**Comment by Fryguy** [comment]
@PanSpagetka This looks great!  A couple of questions...

- What does the "Upload" button do?  Is that for the "core" (non-provider) data collection?  If so, that's not clear to me.
- I don't think we should show the type column as that's an insider thing.  Perhaps instead show `.class.description` (e.g. for vmware it would show "VMware vCenter")?

File: app/assets/config/miq_plugin_example_manifest.js [hunk 11]
```diff
@@ -0,0 +1,2 @@
+//= link_directory ../javascripts/red_hat_cloud_services .js
```
Comment (Fryguy) [inline app/assets/config/miq_plugin_example_manifest.js]: Is this file needed? The file name is plugin_example.

File: config/routes.rb [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#RedHatCloudServices::Engine.routes.draw do
```
Comment (Fryguy) [inline config/routes.rb]: Can you remove this commented line?

**Comment by martinpovolny** [comment]
> * What does the "Upload" button do?  Is that for the "core" (non-provider) data collection?  If so, that's not clear to me.

According to @h-kataria, this is for uploading **all** data. That is both non-provider and provider. Can you, please, confirm that @h-kataria?

**Comment by martinpovolny** [comment]
> What happens when the user clicks on the synchronize? can we show some kind of status info? it would be helpful to show that.

@terezanovotna:  One time data upload will be attempted.

As far as I know, there's no status info (yet) on the backend. Can you confirm that @Fryguy ?

**Comment by h-kataria** [comment]
> > * What does the "Upload" button do?  Is that for the "core" (non-provider) data collection?  If so, that's not clear to me.
> 
> According to @h-kataria, this is for uploading **all** data. That is both non-provider and provider. Can you, please, confirm that @h-kataria?

@Fryguy "Upload" button will upload ALL MIQ data to Cloud and "Synchronize" button will only upload data for selected Providers. That's what was discussed with you, @gtanzillo and @Loicavanel during our conversation in Mahwah.

**Comment by PanSpagetka** [comment]
API doesn't return any human readable type. `.class.description` is not accessible (or at least I don't know how), but I have added second API call so I can convert it. Some providers still don't have description, but I think that they shouldn't be present in the list (mostly Network Managers).

Screenshot:
Screenshot from 2019-07-01 11-49-43

**Comment by terezanovotna** [comment]
@PanSpagetka @h-kataria @martinpovolny and Loic discussed this on BJ today. 

Here is an updated version of how UI should look like. Please share any feedback in the comments to help us clarify what we have as Robin is implementing it. 

https://marvelapp.com/i6c3d3j/screen/58779044

@Fryguy When a user clicks on Synchronize button, we need some kind of feedback loop to tell him what is happening. Do we have that information?

For example:
 - if it's loading 45 seconds, we want to tell the user it is being synchronized - with a spinner
 - when clicked on the synchonize button, do we have any status info that we can share with the user?

@Fryguy thoughts? what's on your roadmap regarding the feedback loop?

**Comment by PanSpagetka** [comment]
I have implemented new design with Loic's comment. When user click on synchronize button, it shows flash message.

Screenshots:
Screenshot from 2019-07-02 14-48-37
Screenshot from 2019-07-02 14-47-43
Screenshot from 2019-07-02 14-47-57

**Comment by h-kataria** [comment]
@Loicavenel @terezanovotna cc

**Comment by martinpovolny** [comment]
@h-kataria: Given this whole repo is a work in progress I think we can merge this and do follow up work in a separate PR or PRs.

**Comment by Loicavenel** [comment]
@h-kataria Platform Synchronization is confusing. Let's call it "Global Synchronization"

File: app/assets/config/miq_plugin_example_manifest.js (hunk 11 above)
Comment (martinpovolny) [reply app/assets/config/miq_plugin_example_manifest.js]: @PanSpagetka : can you remove this?

**Comment by h-kataria** [comment]
@martinpovolny do you have merge rights on this repo, if you have them, let's merge this PR. @PanSpagetka can create a follow up PR to make other changes.

**Comment by Fryguy** [comment]
+1 on merge and follow up.  Also, I'll get permissions sorted out after this.

File: app/assets/config/miq_plugin_example_manifest.js (hunk 11 above)
Comment (Fryguy) [reply app/assets/config/miq_plugin_example_manifest.js]: @PanSpagetka Please also do this in your followup

**Comment by martinpovolny** [comment]
Yes, I have the merge perms, but did not want to merge it w/o your say ;-)

@Fryguy: can you, please, give the perms to @h-kataria too?

**Comment by terezanovotna** [comment]
@inoxx03 is going to help us with wording!
